C error found - error: '_option_main__Integer' undeclared

### Describe the bug

```shell
>v run linkedlist2.v
==================
C:/Users/Administrator/AppData/Local/Temp/v_0/linkedlist2.01HX1NGJ2EW8P371KYV3GZRVH1.tmp.c:16057: warning: assignment from incompatible pointer type
C:/Users/Administrator/AppData/Local/Temp/v_0/linkedlist2.01HX1NGJ2EW8P371KYV3GZRVH1.tmp.c:16100: error: '_option_main__Integer' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Reproduction Steps

# V source code
```v
import time
// 41s.
struct Node<T> {
  mut:
    data T
    prev ?&Node<T>
    next ?&Node<T>
}

struct LinkedList<T> {
  mut:
    size usize
    head ?&Node<T>
}

fn new_linked_list[T]() &LinkedList<T> {
  return &LinkedList<T>{}
}

fn (mut li LinkedList<T>) add[T](data T) {
  mut node := &Node<T>{ data, none, none }
  
  if li.head == none {
    li.head = node
    node.next = node
    node.prev = node
  } else {
    node.next = li.head
    node.prev = li.head?.prev
    node.prev?.next = node
    li.head?.prev = node
  }
  
  li.size += 1
}

fn (mut li LinkedList<T>) pop[T]() ?T {
  if li.head == none {
    return none
  }
  if li.size == 1 {
    data := li.head?.data
    li.head?.next = none
    li.head?.prev = none
    li.head = none
    li.size -= 1
    return data
  }
  
  mut tail := li.head?.prev?
  mut curr := tail.prev?
  curr.next = li.head
  li.head?.prev = curr
  
  tail.next = none
  tail.prev = none
  li.size -= 1
  
  return tail.data
}

@[heap]
struct Integer {
  value int
}

fn main() {
max_itr := 10
t := time.now()
for itr in 0 .. max_itr {
  mut list := new_linked_list<&Integer>()
  println('Itr#$itr list size: $list.size')
  
  list.add(&Integer{10})
  println('Itr#$itr list size: $list.size')
  list.add(&Integer{20})
  println('Itr#$itr list size: $list.size')
  
  n := list.pop()
  println('Itr#$itr list size: $list.size, data: ${n?}')
  m := list.pop()
  println('Itr#$itr list size: $list.size, data: ${m?}')
  o := list.pop()
  println('Itr#$itr list size: $list.size, data: $o')
  
  max := 1000_0000
  step := 100_0000
  for i in 0 .. max {
    list.add(&Integer{i})
    if i % step == 0 {
      println('Itr#$itr list add: $list.size, i: $i')
    }
  }
  for i in 0 .. max {
    list.pop()
    if i % step == 0 {
      println('Itr#$itr list pop: $list.size, i: $i')
    }
  }
  println('Itr#$itr list size: $list.size')
}
  d := time.since(t)
  println('Bye(time $d)!')
}
```

# Run it
```shell
>v run linkedlist2.v
```

### Expected Behavior

Run normally.

### Current Behavior

C error found.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.5 0459519

### Environment details (OS name and version, etc.)

```shell
>v doctor
V full version: V 0.4.5 937a36a.0459519
OS: windows, Microsoft Windows 10 רҵ v19042 64 λ
Processor: 4 cpus, 64bit, little endian,

getwd: D:\code\lang\vlang\ds
vexe: D:\Programs\v\v.exe
vexe mtime: 2024-05-02 06:28:06

vroot: OK, value: D:\Programs\v
VMODULES: OK, value: C:\Users\Administrator\.vmodules
VTMP: OK, value: C:\Users\Administrator\AppData\Local\Temp\v_0

Git version: git version 2.33.1.windows.1
Git vroot status: Error: fatal: not a git repository (or any of the parent directories): .git
.git/config present: false

CC version: Error: 'cc' is not recognized as an internal or external command,
operable program or batch file.

thirdparty/tcc: N/A

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6989,7 +6989,7 @@ fn (mut g Gen) or_block(var_name string, or_block ast.OrExpr, return_type ast.Ty
 			if g.fn_decl.return_type == ast.void_type {
 				g.writeln('\treturn;')
 			} else {
-				styp := g.typ(g.fn_decl.return_type)
+				styp := g.typ(g.fn_decl.return_type).replace('*', '_ptr')
 				err_obj := g.new_tmp_var()
 				g.writeln('\t${styp} ${err_obj};')
 				g.writeln('\tmemcpy(&${err_obj}, &${cvar_name}, sizeof(_option));')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4901,7 +4901,7 @@ fn (mut g Gen) ident(node ast.Ident) {
 	g.write(g.get_ternary_name(name))
 	if is_auto_heap {
 		g.write('))')
-		if is_option {
+		if is_option && node.or_expr.kind != .absent {
 			g.write('.data')
 		}
 	}

--- a/vlib/v/tests/option_ptr_unwrap_test.v
+++ b/vlib/v/tests/option_ptr_unwrap_test.v
@@ -1,0 +1,89 @@
+import time
+// 41s.
+
+struct Node[T] {
+mut:
+	data T
+	prev ?&Node[T]
+	next ?&Node[T]
+}
+
+struct LinkedList[T] {
+mut:
+	size usize
+	head ?&Node[T]
+}
+
+fn new_linked_list[T]() &LinkedList[T] {
+	return &LinkedList[T]{}
+}
+
+fn (mut li LinkedList[T]) add[T](data T) {
+	mut node := &Node[T]{data, none, none}
+
+	if li.head == none {
+		li.head = node
+		node.next = node
+		node.prev = node
+	} else {
+		node.next = li.head
+		node.prev = li.head?.prev
+		node.prev?.next = node
+		li.head?.prev = node
+	}
+
+	li.size += 1
+}
+
+fn (mut li LinkedList[T]) pop[T]() ?T {
+	if li.head == none {
+		return none
+	}
+	if li.size == 1 {
+		data := li.head?.data
+		li.head?.next = none
+		li.head?.prev = none
+		li.head = none
+		li.size -= 1
+		return data
+	}
+
+	mut tail := li.head?.prev?
+	mut curr := tail.prev?
+	curr.next = li.head
+	li.head?.prev = curr
+
+	tail.next = none
+	tail.prev = none
+	li.size -= 1
+
+	return tail.data
+}
+
+@[heap]
+struct Integer {
+	value int
+}
+
+fn test_main() {
+	max_itr := 2
+	t := time.now()
+	for itr in 0 .. max_itr {
+		mut list := new_linked_list[&Integer]()
+		println('Itr#${itr} list size: ${list.size}')
+
+		list.add(&Integer{10})
+		println('Itr#${itr} list size: ${list.size}')
+		list.add(&Integer{20})
+		println('Itr#${itr} list size: ${list.size}')
+
+		mut n := list.pop()
+		println('Itr#${itr} list size: ${list.size}, data: ${n?}')
+		n = list.pop()
+		println('Itr#${itr} list size: ${list.size}, data: ${n?}')
+		n = list.pop()
+		println('Itr#${itr} list size: ${list.size}, data: ${n}')
+	}
+	d := time.since(t)
+	println('Bye(time ${d})!')
+}

